crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,618 @@
+    ///
+    /// In other words, if any component from `B1` is present, then all of the components from `B2` must be.
+    #[inline]
+    pub fn requires_all() -> Self {
+        Self {
+            premise: ArchetypeStatement::<B1>::any_of(),
+            consequence: ArchetypeStatement::<B2>::all_of(),
+        }
+    }
+
+    /// Creates an archetype invariant where any components of `B1` must appear with some components of `B2`.
+    ///
+    /// In other words, if any component from `B1` is present, then at least one component from `B2` must be.
+    #[inline]
+    pub fn requires_one() -> Self {
+        Self {
+            premise: ArchetypeStatement::<B1>::any_of(),
+            consequence: ArchetypeStatement::<B2>::any_of(),
+        }
+    }
+}
+
+impl<B: Bundle> ArchetypeInvariant<B, B> {
+    /// Creates an archetype invariant where all components of `B` require each other.
+    ///
+    /// In other words, if any component of this bundle is present, then all of them must be.
+    #[inline]
+    pub fn atomic() -> Self {
+        Self {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B>::all_of(),
+        }
+    }
+
+    /// Creates an archetype where components of `B` cannot appear with each other.
+    ///
+    /// In other words, if any component of this bundle is present, then no others can be.
+    /// This is particularly useful for creating enum-like groups of components, such as `Dead` and `Alive`.
+    #[inline]
+    pub fn disjoint() -> Self {
+        Self {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B>::at_most_one_of(),
+        }
+    }
+
+    /// Creates an archetype invariant where components of `B` can only appear with each other.
+    ///
+    /// In other words, if any component of this bundle is present, then _only_ components from this bundle can be present.
+    #[inline]
+    pub fn exclusive() -> Self {
+        Self {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B>::only(),
+        }
+    }
+}
+
+/// A statement about the presence or absence of some subset of components in the given [`Bundle`]
+///
+/// This type is used as part of an [`ArchetypeInvariant`].
+///
+/// When used as a premise, the archetype invariant matches all entities which satisfy the statement.
+/// When used as a consequence, then the statment must be true for all entities that were matched by the premise.
+///
+/// For the statements about a single component `C`, wrap it in a single-component bundle `(C,)`.
+/// For single component bundles, `AllOf` and `AnyOf` are equivalent.
+/// Prefer `ArchetypeStatement::<(C,)>::all_of` over `ArchetypeStatement::<(C,)>::any_of` for consistency and clarity.
+///
+/// Note that this is converted to an [`UntypedArchetypeStatement`] when added to a [`World`].
+/// This is to ensure compatibility between different invariants.
```

Comment:
Are these two lines useful to know for end users? They look like an implementation detail.

Reply:
End users will (eventually) want to use an `UntypedArchetypeStatement` in their own code. This is a useful bread crumb, although I see your point and think the framing could be improved.

Reply:
Ah, I hadn't considered the auto-generated statement. Totally agreed- this should be just removed. Fixed as of 22b6af1e36a4ea5a54982c3ddc949212cd18eb2f

Reply:
Agreed. What would you suggest?

Reply:
Cleaned up the docs for this with some help from @alice-i-cecile. Fixed as of 4109138d23ee099dfaab683c61b9c2be117c074b